### DISCUSSION TITLE
Is it possible to unit test Render() method?

### DISCUSSION THREAD

**Discussion by rabbitism** [AvaloniaUI/Avalonia#14786] [Q&A] [answered]
I run a test coverage on my project and Render method is not covered at all. is it possible to test this methods?

**Comment by timunie** [comment, marked as answer]
I think in our UnitTests we render something, save the output and compare it with what was expected. https://github.com/AvaloniaUI/Avalonia/tree/master/tests/Avalonia.RenderTests

**Reply by rabbitism** [reply]
Going back from this, we found it's hard to replicate the tests in this repo because there are so many implementations of internal interface/members, it is not suitable for third parties like me.

**Reply by maxkatz6** [reply]
Don't use avalonia internal tests as a reference.
Use headless platform.

**Reply by maxkatz6** [reply]
https://docs.avaloniaui.net/docs/concepts/headless/
https://github.com/AvaloniaUI/Avalonia.Samples?tab=readme-ov-file#%EF%B8%8F-automated-ui-testing
